### DISCUSSION TITLE
Doesn't work on Meta Quest devices

### DISCUSSION THREAD

**Discussion by FluorescentHallucinogen** [google/model-viewer#3969] [Q&A]
As announced at Meta Connect 2022, Meta Quest Browser has added support for WebXR mixed reality services on Meta Quest 2 and Meta Quest Pro headsets. See https://developer.oculus.com/documentation/web/webxr-mixed-reality/ for more info.

Demos https://immersive-web.github.io/webxr-samples/immersive-ar-session.html and https://aframe-light-and-shadow.glitch.me work like a charm in Meta Quest Browser on my Meta Quest 2 headset.

But https://modelviewer.dev/examples/augmentedreality/ demos unfortunately don't work.

**Steps to reproduce:**

1. Open https://modelviewer.dev/examples/augmentedreality/ in Meta Quest Browser.
2. Click "View in your space" button.

**Result:**
1. The "View in your space" button just disappears after clicking.
2. The following errors are printed in the Meta Quest Browser console (I've used Chrome DevTools Remote debug to see the logs):

```
model-viewer.js:56006 Attempting to present in AR with WebXR...
model-viewer.js:47888 Feature 'hit-test' is not permitted by permissions policy
resolveARSession @ model-viewer.js:47888
present @ model-viewer.js:47948
await in present (async)
[enterARWithWebXR] @ model-viewer.js:56012
await in [enterARWithWebXR] (async)
activateAR @ model-viewer.js:55952
<computed> @ model-viewer.js:55882
model-viewer.js:47888 Feature 'dom-overlay' is not supported for mode: immersive-ar
resolveARSession @ model-viewer.js:47888
present @ model-viewer.js:47948
await in present (async)
[enterARWithWebXR] @ model-viewer.js:56012
await in [enterARWithWebXR] (async)
activateAR @ model-viewer.js:55952
<computed> @ model-viewer.js:55882
model-viewer.js:47888 Feature 'light-estimation' is not supported for mode: immersive-ar
resolveARSession @ model-viewer.js:47888
present @ model-viewer.js:47948
await in present (async)
[enterARWithWebXR] @ model-viewer.js:56012
await in [enterARWithWebXR] (async)
activateAR @ model-viewer.js:55952
<computed> @ model-viewer.js:55882
model-viewer.js:56015 Error while trying to present in AR with WebXR
[enterARWithWebXR] @ model-viewer.js:56015
await in [enterARWithWebXR] (async)
activateAR @ model-viewer.js:55952
<computed> @ model-viewer.js:55882
model-viewer.js:56016 DOMException: Failed to execute 'requestSession' on 'XRSystem': The session request contains invalid requiredFeatures and could not be fullfilled.
    at iu.resolveARSession (https://modelviewer.dev/node_modules/@google/model-viewer/dist/model-viewer.min.js:615:2125)
    at iu.present (https://modelviewer.dev/node_modules/@google/model-viewer/dist/model-viewer.min.js:615:3289)
    at async [enterARWithWebXR] (https://modelviewer.dev/node_modules/@google/model-viewer/dist/model-viewer.min.js:1125:20592)
    at async s.activateAR (https://modelviewer.dev/node_modules/@google/model-viewer/dist/model-viewer.min.js:1125:19539)
[enterARWithWebXR] @ model-viewer.js:56016
await in [enterARWithWebXR] (async)
activateAR @ model-viewer.js:55952
<computed> @ model-viewer.js:55882
model-viewer.js:47998 Error while trying to end WebXR AR session
stopPresenting @ model-viewer.js:47998
[enterARWithWebXR] @ model-viewer.js:56017
await in [enterARWithWebXR] (async)
activateAR @ model-viewer.js:55952
<computed> @ model-viewer.js:55882
model-viewer.js:47999 TypeError: Cannot read properties of null (reading 'end')
    at iu.stopPresenting (model-viewer.js:47994:39)
    at [enterARWithWebXR] (model-viewer.js:56017:50)
    at async s.activateAR (model-viewer.js:55951:22)
stopPresenting @ model-viewer.js:47999
[enterARWithWebXR] @ model-viewer.js:56017
await in [enterARWithWebXR] (async)
activateAR @ model-viewer.js:55952
<computed> @ model-viewer.js:55882
model-viewer.js:56019 Falling back to next ar-mode
[enterARWithWebXR] @ model-viewer.js:56019
await in [enterARWithWebXR] (async)
activateAR @ model-viewer.js:55952
<computed> @ model-viewer.js:55882
model-viewer.js:55958 No AR Mode can be activated. This is probably due to missing configuration or device capabilities
activateAR @ model-viewer.js:55958
[enterARWithWebXR] @ model-viewer.js:56021
await in [enterARWithWebXR] (async)
activateAR @ model-viewer.js:55952
<computed> @ model-viewer.js:55882
```

**Comment by FluorescentHallucinogen** [comment]
If I understand it correctly, the problem is in the `'hit-test'`, `'dom-overlay'`, and `'light-estimation'` features.

But `'hit-test'` is the only required feature (that produces error), `'dom-overlay'` and `'light-estimation'` are optional features (that produce warnings).

And feature `'hit-test'` is not permitted by permissions policy.

**Reply by elalish** [reply]
The real issue is that we support AR via WebXR, but not VR. There have been open issues on MV for a long time regarding adding VR support, but it's a whole different UX paradigm and we don't even have a designer for it. If you'd like to come up with something cool for VR, a PR would be most welcome. I don't anticipate having time for it myself in the foreseeable future.

**Reply by FluorescentHallucinogen** [reply]
> The real issue is that we support AR via WebXR, but not VR.

But Meta Quest devices support AR via WebXR since October 2022. At least https://immersive-web.github.io/webxr-samples/immersive-ar-session.html and https://aframe-light-and-shadow.glitch.me work fine.

It seems the real issue is that `'hit-test'` permissions policy is not enabled for https://modelviewer.dev origin. See https://developer.chrome.com/docs/privacy-sandbox/permissions-policy/ for more details.

**Reply by elalish** [reply]
Ah, interesting. I can't set that permission from my side; generally that's an iframe thing. Perhaps this is a bug in their browser?

**Reply by FluorescentHallucinogen** [reply]
Here are the permissions (features) currently enabled:

![screenshot](https://user-images.githubusercontent.com/7892779/203608924-e0edf6b2-efbd-4d82-a5ae-2a01aa63a54e.png)

You can check that yourself. See https://developer.chrome.com/docs/privacy-sandbox/permissions-policy/#chrome-devtools-integration.

Could we try to enable the `'hit-test'` too?

**Reply by FluorescentHallucinogen** [reply]
The same problem on https://immersive-web.github.io/webxr-samples/hit-test.html.

**Reply by FluorescentHallucinogen** [reply]
According to https://immersive-web.github.io/hit-test/#hit-test-feature-descriptor, the `'hit-test'` feature requires `"xr-spatial-tracking"` policy, but it's allowed. 🤯

This seems to be a bug in the Meta Quest Browser.

**Reply by FluorescentHallucinogen** [reply]
@elalish As an idea of workaround, is it possible to somehow disable hit test for `<model-viewer>` (e.g. via property or attribute)? What if I want to just view models without placing on real-world surfaces?

**Reply by elalish** [reply]
No, that would break the experience (and the code). Let's just push Meta to fix their bug.

**Reply by FluorescentHallucinogen** [reply]
@davehill00 @cabanier @pampas93 Could you please take a look?

**Reply by cabanier** [reply]
The Quest browser does not support the hit test feature.
An experience can fall back to planes or just the floor.

**Reply by FluorescentHallucinogen** [reply]
@elalish
> No, that would break the experience (and the code). Let's just push Meta to fix their bug.

As it turns out, it's not a bug, hit test is simply not supported in Meta Quest Browser yet.

I don't understand in what way that break the experience? And even more, I don't understand how that can break the code?

I've proposed to implement a new option (e.g. property or attribute for `<model-viewer>` element) to disable hit test. By default, this option will be disabled. So nothing changes for devices that support hit test. But this option will allow to view 3D models on Meta Quest devices that currently doesn't support hit test. It's better than no support. So this will improve the experience, not break it.

Moreover, ideally the option is not needed at all. The `<model-viewer>` should use feature detection and graceful depredate or progressively enhance.

I played around with https://aframe-light-and-shadow.glitch.me demo that doesn't use hit test in my Meta Quest 2 headset and I can't say that the experience is bad. The 3D model is just placed on the floor. But I can view it from all angles.

@cabanier Any ETA of hit test support in Meta Quest Browser?

**Reply by cabanier** [reply]
> @cabanier Any ETA of hit test support in Meta Quest Browser?

It will never come to Quest 2. The cameras aren't good enough to do real world hit detection.
I considered emulating it if the user ran room setup but that has drawbacks as well (ie few users manually scan their room.)

**Reply by elalish** [reply]
Ah, that's useful info. I had figured if planes were supported then hit-test would simply return the plane hits. @cabanier I'm not used to any hit detection being particularly accurate in WebXR, but we make due; what's your mental model for the distinction between planes and hits?

@FluorescentHallucinogen I'd be supportive of a PR here. It'll complicate the code a bit, but it should be possible to fall back to planes if hit-test isn't available. And I suppose the experience is actually decent with neither since we now allow you to move the object even when hits aren't detected, which means there's probably no need for a configuration. Do you want to give it a try?

**Reply by FluorescentHallucinogen** [reply]
@elalish
> I'd be supportive of a PR here.
> Do you want to give it a try?

Yes, I'll try to create a pull request.

@cabanier
> I considered emulating it if the user ran room setup but that has drawbacks as well (ie few users manually scan their room.)

Could you please clarify what you mean, will the Meta Quest Browser emulate the hit test itself?

The internet is full of demos that have `'hit-test'` in `requiredFeatures`, something like:

```js
await navigator.xr!.requestSession!('immersive-ar', {
  requiredFeatures: ['hit-test'],
  optionalFeatures: ['dom-overlay', 'light-estimation'],
});
```

And it would be very nice if such demos would work on Meta Quest devices without any changes in the code.

**Reply by cabanier** [reply]
> .. what's your mental model for the distinction between planes and hits?

Planes are supported natively on our platform but not hit testing.
I see that Microsoft did something like you suggest for Hololens but I'm hesitant to go that way; mostly because almost nobody runs room setup so hit test wouldn't  "hit" anything. This would be a bad experience for our users...

**Reply by cabanier** [reply]
> And it would be very nice if such demos would work on Meta Quest devices without any changes in the code.

I have not seen a lot of content that requires hit testing but not dom overlay. Can you point to some examples?

**Reply by davehill00** [reply]
> > .. what's your mental model for the distinction between planes and hits?
> 
> Planes are supported natively on our platform but not hit testing. I see that Microsoft did something like you suggest for Hololens but I'm hesitant to go that way; mostly because almost nobody runs room setup so hit test wouldn't "hit" anything. This would be a bad experience for our users...

One option would be to use the "ground plane" as a fallback (i.e., whatever the ground plane is from Guardian setup) and then use the Room Planes if they exist.

For model-viewer in particular, we've talked about adding an API that will let a website prompt the user to run room setup. We could incorporate that into the flow if model-viewer detects that planes are not set up.

**Reply by cabanier** [reply]
> One option would be to use the "ground plane" as a fallback (i.e., whatever the ground plane is from Guardian setup) and then use the Room Planes if they exist.

The ground plane is almost always known so an experience could make hit testing optional and fall back to local-floor

> For in particular, we've talked about adding an API that will let a website prompt the user to run room setup. We could incorporate that into the flow if detects that planes are not set up.

Yes. We shipped that API as part of WebXR Planes.

**Reply by FluorescentHallucinogen** [reply]
@cabanier @davehill00

> almost nobody runs room setup so hit test wouldn't "hit" anything. This would be a bad experience for our users...

E.g. https://spatialfusion.io/ displays a custom dialog warning that the user needs to set up a room.

But native browser warning would be much better.

> For in particular, we've talked about adding an API that will let a website prompt the user to run room setup. We could incorporate that into the flow if detects that planes are not set up.

> Yes. We shipped that API as part of WebXR Planes.

Could you please share the API docs how to show the request to set up a room?

**Reply by cabanier** [reply]
> > Yes. We shipped that API as part of WebXR Planes.
> 
> Could you please share the API docs how to show the request to set up a room?

https://immersive-web.github.io/real-world-geometry/plane-detection.html#dom-xrsession-initiateroomcapture

**Reply by FluorescentHallucinogen** [reply]
@elalish Why `'hit-test'` is required feature in `<model-viewer>` and not optional as in A-Frame (https://aframe.io/docs/1.3.0/components/ar-hit-test.html)?

**Reply by elalish** [reply]
@FluorescentHallucinogen Honestly, a lot of changes have happened since `hit-test` was added. This may be as simple as making it optional now; give it a try and see how the experience is. 

@cabanier For phone AR, local-floor is basically useless, hence we ignored it. Does it give a reasonable result on Quest? I'm curious how one should properly implement the fallback behaviors to handle these different combinations.

I was confused about planes; in phone AR these are created from the same data that creates the hit-test results (camera). I didn't realize Quest had a separate flow for static planes. I see the distinction now. I was actually thinking hit-test could return hits from the local-floor if that's reasonably accurate, so that it would work without room setup. Then at least the fallbacks would be taken care of for me 😁.

**Reply by cabanier** [reply]
> @cabanier For phone AR, local-floor is basically useless, hence we ignored it. Does it give a reasonable result on Quest? I'm curious how one should properly implement the fallback behaviors to handle these different combinations.

Yes. As long as the user defined the floor correctly. I suspect that's correct in most cases because otherwise most games wouldn't function correctly.

> I was confused about planes; in phone AR these are created from the same data that creates the hit-test results (camera). I didn't realize Quest had a separate flow for static planes. I see the distinction now. I was actually thinking hit-test could return hits from the local-floor if that's reasonably accurate, so that it would work without room setup. Then at least the fallbacks would be taken care of for me 😁.

Quest only has access to the planes that were created by room setup and the floor planes.
I agree that adding support for local-floor on your side will be the quickest and easiest solution. The hit test code is quite complex because it tests all hits against all planes continuously.

**Reply by FluorescentHallucinogen** [reply]
@elalish In case you're wondering what WebXR AR/MR models (scenes) look like on Meta Quest devices, check out https://web.dev/pwas-on-oculus-2/#augmentedmixed-reality-passthrough. :wink:

**Reply by cabanier** [reply]
Hit testing will be in the next version of the Quest browser. Apart from a [bug in the framework](https://github.com/google/model-viewer/pull/3984) this should unblock model-viewer on our platform.

**Reply by FluorescentHallucinogen** [reply]
Uh, I can't wait to see it alive on my headset! 🤯

@cabanier Any ETA of the next version of the Meta Quest Browser? BTW, do you have Beta and Canary release channels? And do you have a public bug tracker?

@elalish Is https://github.com/google/model-viewer/pull/3984 ready to merge? :-/

**Reply by cabanier** [reply]
Early 2023. We don't have public beta release channels :-\

**Reply by FluorescentHallucinogen** [reply]
@cabanier
> Hit testing will be in the next version of the Quest browser.
> Early 2023.

In what version number exactly? Does it also require a v48 firmware update?

**Reply by cabanier** [reply]
25.2 or above. It doesn't require a firmware update.

**Reply by FluorescentHallucinogen** [reply]
Just checked in the new version of Meta Quest Browser. It works! 🎉

@cabanier Thank you so much for your work! :clap:

Some nits:

Some models fall to the floor in ~1-2 seconds, but some for some reason hover in the air. I've tried in both, Roomscale Boundary and Stationary Boundary.

You can check it yourself on https://modelviewer.dev/examples/augmentedreality/index.html.

The first model (chair) hovers in the air, but the second (mixer) and third (cactus) fall to the floor. Why the first model (chair) doesn't fall to the floor too? Is it because of its size?

BTW, I have added Desk and Couch into my boundary. Why are these planes not used?

Also, at the moment, it is not possible to move, rotate and zoom models.

I will try to add support for controllers and hand tracking for this. :wink:

**Reply by cabanier** [reply]
> Some models fall to the floor in ~1-2 seconds, but some for some reason hover in the air. I've tried in both, Roomscale Boundary and Stationary Boundary.

Can you try https://threejs.org/examples/webxr_ar_hittest and see if the reticle is hitting your planes?
If you marked up your room, you should be able to see that.

Let me know if it doesn't work and I will check things on my side

> BTW, I have added Desk and Couch into my boundary. Why are these planes not used?
> Also, at the moment, it is not possible to move, rotate and zoom models.
> I will try to add support for controllers and hand tracking for this. 😉

Thanks! Will it be possible to manipulate the model once you do so?
Let me know if you need any help!

**Reply by FluorescentHallucinogen** [reply]
@cabanier
> Let me know if you need any help!

Do you know any good WebXR examples/demos to move, rotate and zoom objects with controllers and/or hands?

**Reply by cabanier** [reply]
> Do you know any good WebXR examples/demos to move, rotate and zoom objects with controllers and/or hands?

Are you looking for a site or an open source example? [Plockle](https://plockle.com/)  is very good but not public.
I see that he has an [open source library](https://plockle.com/) for hand interaction.

Three.js has a bunch of examples (ie https://threejs.org/examples/webxr_vr_dragging or https://threejs.org/examples/webxr_vr_handinput_cubes) that have simple hand or controllers interactions with geometries.

**Reply by FluorescentHallucinogen** [reply]
Since the objects are on the floor and can be very large (closet, car, etc.), taking them in hands may not be a good UX.

Also need to figure out how to resize objects.

So I'm looking for something like what Windows Mixed Reality calls a "Bounding box", but for WebXR:

https://user-images.githubusercontent.com/7892779/214134467-d976a517-2a96-42bb-8bd5-bff9f06d65a1.mp4

![hololens2_boundingbox_rotate](https://user-images.githubusercontent.com/7892779/214138542-4f045f3b-4852-479c-b820-701c06b1ba28.gif)

![hololens2_boundingbox_resize](https://user-images.githubusercontent.com/7892779/214138563-a1fdcaf0-faeb-46b1-bd20-d6b6cba6a595.gif)

**Reply by FluorescentHallucinogen** [reply]
Two handed move, scale and rotate:

![two-handed](https://user-images.githubusercontent.com/7892779/214144814-ece1f3af-02a1-4ef7-a814-195a6cfe9469.gif)

**Reply by FluorescentHallucinogen** [reply]
@cabanier After some additional research I've found that only Babylon.js supports the bounding box gizmo.

![gizmos](https://user-images.githubusercontent.com/7892779/216830463-1caf85e3-6607-4bd0-8015-d8c05b1ef4c4.png)

You can play with it here: https://playground.babylonjs.com/#8GY6J8#211. It supports Quest controllers and hands. :wink:

Three.js only has `TransformControls` (similar to Blender), which has 3 modes: `translate`, `rotate` and `scale`. And the problems are that:

1. it's not possible to enable more than one or all modes at the same time
2. it's not intuitive for end users

You can play with it here: https://threejs.org/examples/#misc_controls_transform.

Ideally, the `<model-viewer>` should have something like this (with arrow hints and size/rotation value hints):

![1_xh8lnfJrgS49aSMMV5I1sw](https://user-images.githubusercontent.com/7892779/216831623-f780b5f3-125e-47aa-bc19-6e9041911bdb.png)

https://user-images.githubusercontent.com/7892779/216831883-8211c3ff-3d96-4fb1-bd82-583cb3e3eb3c.mp4

https://user-images.githubusercontent.com/7892779/216831889-7c5e63e6-302f-466e-8528-043df02364e8.mp4

Taken from this awesome blog post: https://medium.com/@alexia.buclet/ux-poc-in-ar-vr-2-manipulation-gizmo-95ad8c5c5580.

See also https://youtu.be/3Mk7Vgcv1jQ. :wink:

Something very similar is implemented in Meta Horizon Worlds. And I've heard that Meta working on the web version of Horizon Worlds.

@cabanier Do you happen to know if it will use Three.js under the hood? If so, is there any chance that Meta will release some kind of object management helper library for web?

@elalish Should this be implemented on the `<model-viewer>` side or on the Three.js side?

**Reply by elalish** [reply]
I'd recommend Three.js - seems generically useful. You'll also get better feedback from their larger community.

**Reply by cabanier** [reply]
> I'd recommend Three.js - seems generically useful. You'll also get better feedback from their larger community.

Yes, **model-viewer** is designed on top of three.js so it's unlikely that we will pull in another framework for this feature.

**Reply by FluorescentHallucinogen** [reply]
I've opened https://github.com/mrdoob/three.js/issues/25619.

**Reply by FluorescentHallucinogen** [reply]
@elalish @cabanier I've implemented a simpler (without bounding box transform controls) solution to the problem. PTAL at #4643. ;)

The code is based on https://github.com/mrdoob/three.js/blob/master/examples/webxr_xr_dragging.html.

It works! :)

But I need some help with:

1. For some reason the controller models are small and are not in the right place, but far away from me, next to the 3D model. See https://youtu.be/PXChXTSLBJE. I assume it's because of the scene resizing.
2. I've used `any` and `@ts-ignore` in multiple places in the code. Trying to use the correct types leads to more and more TypeScript errors. :(

The pull request allows edits by maintainers.

After that, I plan to implement model resizing using two controllers/hands. ;)

**Reply by cabanier** [reply]
> For some reason the controller models are small and are not in the right place, but far away from me, next to the 3D model. See https://youtu.be/PXChXTSLBJE. I assume it's because of the scene resizing.

I agree. There must be an additional resize. Are you sure your controller object are attached on the top node? Is there scaling applied to that node?
